### PR TITLE
script: Fix thread-safety issue in python interpreter execution

### PR DESCRIPTION
**Pull request by honggyukim**
This PR is to fix thread-safety problem running python script in record time.

Before:
```
$ uftrace record -S ./scripts/replay.py tests/t-thread
# DURATION    TID     FUNCTION
            [30186] | main() {
            [30186] |   pthread_create() {
  35.314 us [30186] |   } /* pthread_create */
            [30186] |   pthread_create() {
  19.590 us [30186] |   } /* pthread_create */
            [30186] |   pthread_create() {
  96.816 us [30186] |   } /* pthread_create */
            [30186] |   pthread_create() {
process crashed by signal 11: Segmentation fault (si_code: 1)
            [30189] | foo() {
child terminated by signal: 11: Segmentation fault
```
After:
```
$ uftrace record -S ./scripts/replay.py tests/t-thread
# DURATION     TID     FUNCTION
            [ 29732] | main() {
            [ 29732] |   pthread_create() {
  29.533 us [ 29732] |   } /* pthread_create */
            [ 29732] |   pthread_create() {
 156.904 us [ 29732] |   } /* pthread_create */
            [ 29732] |   pthread_create() {
  66.442 us [ 29732] |   } /* pthread_create */
            [ 29732] |   pthread_create() {
            [ 29735] | foo() {
            [ 29735] |   a() {
            [ 29735] |     b() {
            [ 29737] | foo() {
            [ 29737] |   a() {
            [ 29737] |     b() {
            [ 29737] |       c() {
   5.265 us [ 29737] |       } /* c */
  21.862 us [ 29737] |     } /* b */
  35.246 us [ 29737] |   } /* a */
  81.177 us [ 29737] | } /* foo */
            [ 29738] | foo() {
            [ 29738] |   a() {
            [ 29738] |     b() {
            [ 29738] |       c() {
   6.112 us [ 29738] |       } /* c */
  25.124 us [ 29738] |     } /* b */
  40.623 us [ 29738] |   } /* a */
 141.049 us [ 29738] | } /* foo */
            [ 29736] | foo() {
            [ 29736] |   a() {
            [ 29736] |     b() {
            [ 29736] |       c() {
   3.384 us [ 29736] |       } /* c */
  14.555 us [ 29736] |     } /* b */
  23.793 us [ 29736] |   } /* a */
 230.809 us [ 29736] | } /* foo */
  20.810 us [ 29732] |   } /* pthread_create */
            [ 29732] |   pthread_join() {
            [ 29735] |       c() {
 252.246 us [ 29735] |       } /* c */
 269.668 us [ 29735] |     } /* b */
 285.085 us [ 29735] |   } /* a */
 327.011 us [ 29735] | } /* foo */
  75.777 us [ 29732] |   } /* pthread_join */
            [ 29732] |   pthread_join() {
   3.518 us [ 29732] |   } /* pthread_join */
            [ 29732] |   pthread_join() {
   4.658 us [ 29732] |   } /* pthread_join */
            [ 29732] |   pthread_join() {
   4.206 us [ 29732] |   } /* pthread_join */
 696.298 us [ 29732] | } /* main */
```

### CODE REVIEW

**Comment by honggyukim** [comment]
Hi @namhyung, I would like to add a flush routine for thread creation but don't know where to put it. Please let me know if you have an idea on it.

I also found strange problem when running gcc and node.  They invoke `mcount_startup` many times. It seems that each thread calls it. Do you have any ideas on it?

**Comment by namhyung** [comment]
Why (and what) do you want to flush on thread creation?

The `mcount_startup` will be called during (module) intialization so it'll be same to the number of sessions.

File: utils/script-python.c [hunk 1]
```diff
@@ -139,13 +139,15 @@ static int set_python_path(char *py_pathname)
 static int import_python_module(char *py_pathname)
 {
 	PyObject *pName;
-	char *py_basename = basename(py_pathname);
+	char *py_basename = xstrdup(basename(py_pathname));
 
 	pr_dbg("%s(%s)\n", __func__, py_pathname);
```
Comment (namhyung) [inline utils/script-python.c]: Could you please show me what it prints before applying this patch?

Comment (honggyukim) [reply utils/script-python.c]: If I don't add this change, the output shows as below:
```
$ uftrace -S scripts/report-libcall.py -v /usr/bin/gcc ~/hello.c
    ...
mcount: initializing mcount library
script: script_init(scripts/report-libcall.py)
script: script_init_for_python("scripts/report-libcall.py")
script: set_python_path("scripts/report-libcall.py")
script: import_python_module("scripts/report-libcall.py")
script: python initialization finished
mcount: mcount setup done
mcount: initializing mcount library
script: script_init(scripts/report-libcall)
scripts/report-libcall: No such file or directory
mcount: mcount setup done
mcount: shmem_finish: tid: 752 seqnum = 18 curr = 0, nr_buf = 2 max_buf = 2
mcount: destroy plthook special function index
mcount: initializing mcount library
script: script_init(scripts/report-libcall)
scripts/report-libcall: No such file or directory
mcount: mcount setup done
mcount: shmem_finish: tid: 753 seqnum = 3 curr = 1, nr_buf = 2 max_buf = 2
mcount: destroy plthook special function index
mcount: initializing mcount library
script: script_init(scripts/report-libcall)
scripts/report-libcall: No such file or directory
mcount: mcount setup done
mcount: initializing mcount library
script: script_init(scripts/report-libcall)
scripts/report-libcall: No such file or directory
mcount: mcount setup done
mcount: shmem_finish: tid: 756 seqnum = 10 curr = 0, nr_buf = 2 max_buf = 2
mcount: destroy plthook special function index
mcount: shmem_finish: tid: 754 seqnum = 0 curr = 0, nr_buf = 2 max_buf = 2
mcount: destroy plthook special function index
child terminated by signal: 11: Segmentation fault
    ...
```

File: scripts/report-libcall.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+def uftrace_exit(ctx):
+    pass
+
+def uftrace_end():
+    global libcall_map
+    sorted_dict = sorted(libcall_map.items(), key=operator.itemgetter(1), reverse=True)
+
+    pid = os.getpid()
+    with open("/proc/%s/comm" % pid) as proc_comm:
```
Comment (honggyukim) [inline scripts/report-libcall.py]: I think if there's a way to identify if the script is execution at record or replay time.  This code only makes sense when it runs at record time.

**Comment by honggyukim** [comment]
The output result of running `report-libcall.py` in record time is as follows:
```
$ uftrace -S scripts/report-libcall.py /usr/bin/gcc hello.c
  # Library Function Call Report for 'cc1' (pid: 3515)

       Call Count    Library Functions
  ================  ================================================
            16327    strlen
            14561    memcmp
             9676    strcmp
             9237    memcpy
             8135    memset
             5341    malloc
    ...
  # Library Function Call Report for 'as' (pid: 3522)

       Call Count    Library Functions
  ================  ================================================
             5221    strcmp
             2742    strlen
             2025    malloc
             2004    xmalloc
              190    memcpy
    ...
  # Library Function Call Report for 'ld' (pid: 3534)

       Call Count    Library Functions
  ================  ================================================
             3722    strpbrk
             3254    bfd_hash_lookup
             2713    bfd_check_format_matches
             2699    bfd_check_format
    ...
  # Library Function Call Report for 'collect2' (pid: 3523)

       Call Count    Library Functions
  ================  ================================================
            17912    getc
            13261    putc
              506    fflush
              504    fputs
    ...
                1    __errno_location
                1    textdomain
                1    vfork


child terminated by signal: 11: Segmentation fault

```

**Comment by namhyung** [comment]
I tried to run the script but it got stuck in a mutex - maybe there's a recursion somewhere.

File: utils/script-python.c (hunk 1 above)
Comment (namhyung) [reply utils/script-python.c]: Ah, this was because the script_pathname is an environment variable.  You fix looks fine then.

Comment (namhyung) [reply utils/script-python.c]: But following debug messages are duplicated so I think only the first one needs level 1.  Please change others to `pr_dbg2` (or simply delete them).

```
script: script_init("scripts/report-libcall.py")
script: script_init_for_python("scripts/report-libcall.py")
script: set_python_path("scripts/report-libcall.py")
script: import_python_module("scripts/report-libcall.py")
script: python initialization finished
```

**Comment by namhyung** [comment]
I was able to fix it with the [below patch.txt](https://github.com/namhyung/uftrace/files/1520142/plthook-fix-recursion.patch.txt).  But then I don't see a segfault.

```diff
diff --git a/libmcount/plthook.c b/libmcount/plthook.c
index 507d985..c51c6dd 100644
--- a/libmcount/plthook.c
+++ b/libmcount/plthook.c
@@ -759,8 +759,12 @@ unsigned long plthook_entry(unsigned long *ret_addr, unsigned long child_idx,
                if (mtdp == NULL)
                        goto out;
        }
-       else
+       else {
+               if (unlikely(mtdp->recursion_guard))
+                       goto out;
+
                mtdp->recursion_guard = true;
+       }
 
        recursion = false;
 
```

**Comment by honggyukim** [comment]
Hmm.. I didn't see recursion before. And I applied your patch but got another segfault as well.
```
$ uftrace record -S scripts/report-libcall.py /usr/bin/gcc hello.c
  # Library Function Call Report for 'cc1' (pid: 105701)

       Call Count    Library Functions
  ================  ================================================
            14853    memcmp
            10848    memcpy
             9906    strcmp
             8817    memset
        ...
  # Library Function Call Report for 'as' (pid: 105703)

       Call Count    Library Functions
  ================  ================================================
             5209    strcmp
             2604    strlen
             2002    malloc
             1981    xmalloc
        ...
                1    bfd_openw
                1    textdomain
                1    fopen64


process crashed by signal 11: Segmentation fault (si_code: 1)
  # Library Function Call Report for 'ld' (pid: 105716)

       Call Count    Library Functions
  ================  ================================================
             4582    free
             3212    bfd_hash_lookup
             2681    strcmp
             2673    bfd_check_format_matches
        ...
  # Library Function Call Report for 'collect2' (pid: 105714)

       Call Count    Library Functions
  ================  ================================================
            17482    getc
            12981    putc
              494    fflush
              492    fputs
        ...
                1    __errno_location
                1    textdomain
                1    vfork


child terminated by signal: 11: Segmentation fault
```

**Comment by namhyung** [comment]
Could you please show me any of backtrace?

**Comment by honggyukim** [comment]
Hmm.. There must be some problem on imported modules inside the script.  If I don't import `operator`, everything are running fine.
```
diff --git a/scripts/report-libcall.py b/scripts/report-libcall.py
index 15b3569..ba01ebe 100644
--- a/scripts/report-libcall.py
+++ b/scripts/report-libcall.py
@@ -5,7 +5,6 @@
 #

 import os
-import operator

 libcall_map = {}

@@ -24,7 +23,7 @@ def uftrace_exit(ctx):

 def uftrace_end():
     global libcall_map
-    sorted_dict = sorted(libcall_map.items(), key=operator.itemgetter(1), reverse=True)
+    sorted_dict = sorted(libcall_map.iteritems(), key=lambda (k, v): (v, k), reverse=True)

     pid = os.getpid()
     with open("/proc/%s/comm" % pid) as proc_comm:
```
I need to think about the problem more but `report-libcall.py` works fine with this fix now.

**Comment by honggyukim** [comment]
I changed the `report-libcall.py` and pushed again. It works fine so please take a look at it again!

**Comment by namhyung** [comment]
I tested it and it still has the recursion problem on my Archlinux.  And I got segfault on a Ubuntu 16.04 container.

**Comment by honggyukim** [comment]
Hmm.. my machines including Ubuntu 16.04 and 17.04 work both fine without recursion problem.
I just pushed once again including your recursion fix. Could you please test it again?

**Comment by honggyukim** [comment]
> Could you please show me any of backtrace?

Here is the backtrace.
```
(gdb) bt
#0  update_pltgot (pd=0x0, dyn_idx=0, mtdp=<optimized out>) at /home/honggyu/src/uftrace/git/uftrace/libmcount/plthook.c:707
#1  0x00007f59280d5f06 in plthook_exit (retval=0x7ffdc32362f0) at /home/honggyu/src/uftrace/git/uftrace/libmcount/plthook.c:880
#2  0x00007f59280e532e in plthook_return () at /home/honggyu/src/uftrace/git/uftrace/arch/x86_64/plthook.S:72
#3  0x00007f59280e5319 in plt_hooker () at /home/honggyu/src/uftrace/git/uftrace/arch/x86_64/plthook.S:55
#4  0x000000000154e1f0 in ?? ()
#5  0x0000000100000000 in ?? ()
#6  0x0000000001478810 in ?? ()
#7  0x00000000014b56e0 in ?? ()
#8  0x0000000000000000 in ?? ()
```
And I could see `pd` is NULL.
```
701│ static void update_pltgot(struct mcount_thread_data *mtdp,
702│                           struct plthook_data *pd, int dyn_idx)
703│ {
704│         if (unlikely(plthook_no_pltbind))
705│                 return;
706│
707├───────> if (!pd->resolved_addr[dyn_idx]) {
708│                 unsigned long plthook_addr;
709│ #ifndef SINGLE_THREAD
710│                 static pthread_mutex_t resolver_mutex = PTHREAD_MUTEX_INITIALIZER;
711│
712│                 pthread_mutex_lock(&resolver_mutex);
713│ #endif
714│                 if (!pd->resolved_addr[dyn_idx]) {
715│                         plthook_addr = mcount_arch_plthook_addr(pd, dyn_idx);
716│                         setup_pltgot(pd, 3 + dyn_idx, dyn_idx,
717│                                      (void *)plthook_addr);
718│                 }
719│
720│ #ifndef SINGLE_THREAD
721│                 pthread_mutex_unlock(&resolver_mutex);
722│ #endif
723│         }
724│ }

(gdb) p pd
$1 = (struct plthook_data *) 0x0
```

**Comment by namhyung** [comment]
Yes, I reproduced it.  It's a problem of `vfork` to start `collect2` process.  Strangely, uftrace could not find any exec* for it after vfork so it got a negative rstack idx.

**Comment by namhyung** [comment]
I was able to work around it with the following commit.  But still need to figure out why..

```diff
diff --git a/libmcount/plthook.c b/libmcount/plthook.c
index a0b772e..5efc84e 100644
--- a/libmcount/plthook.c
+++ b/libmcount/plthook.c
@@ -897,7 +897,10 @@ unsigned long plthook_exit(long *retval)
        mtdp->recursion_guard = true;
 
 again:
-       rstack = &mtdp->rstack[mtdp->idx - 1];
+       if (likely(mtdp->idx > 0))
+               rstack = &mtdp->rstack[mtdp->idx - 1];
+       else
+               rstack = restore_vfork(mtdp, NULL);  /* FIXME! */
 
        if (unlikely(rstack->flags & (MCOUNT_FL_LONGJMP | MCOUNT_FL_VFORK))) {
                if (rstack->flags & MCOUNT_FL_LONGJMP) {
```

**Comment by namhyung** [comment]
Sorry for the mistake.  I was able to work around this problem by using vfork-saved rstack for the negative case.  Here is [the commit I used](https://github.com/namhyung/uftrace/files/1525467/plthook-Handle-corrupted-rstack-after-vfork.patch.txt), could you please test it and merge it into your branch if it works?

```diff
diff --git a/libmcount/plthook.c b/libmcount/plthook.c
index a0b772e..5efc84e 100644
--- a/libmcount/plthook.c
+++ b/libmcount/plthook.c
@@ -897,7 +897,10 @@ unsigned long plthook_exit(long *retval)
        mtdp->recursion_guard = true;
 
 again:
-       rstack = &mtdp->rstack[mtdp->idx - 1];
+       if (likely(mtdp->idx > 0))
+               rstack = &mtdp->rstack[mtdp->idx - 1];
+       else
+               rstack = restore_vfork(mtdp, NULL);  /* FIXME! */
 
        if (unlikely(rstack->flags & (MCOUNT_FL_LONGJMP | MCOUNT_FL_VFORK))) {
                if (rstack->flags & MCOUNT_FL_LONGJMP) {
```

**Comment by namhyung** [comment]
It'd be nice to use following commit message for the recursion fix:

> plthook: Handle recursion in plthook_entry()
> 
> There was a deadlock in plthook_entry() due to a recursion.  When
> a (python) script is used, it's possible to call another library
> functions so protect it like mcount_entry() does.
> 
> Signed-off-by: Namhyung Kim <namhyung@gmail.com>

**Comment by honggyukim** [comment]
I've added your patches and updated again. Please take a look at it again. Thanks for your help!

**Pull request merged**